Fix newline issues in new cli project generated files

jonthysell opened the pull request:
* Moves the line ending normalization used by autolinking into the main cli generation, to prevent creating mixed LF and CRLF files
* Updates autolinking code to exactly match new project code, so running autolinking on a fresh cli project produces 0 changes
* Fixed the removal of the clang-format guards from the template files to prevent creating mixed LF and CRLF files

Closes #6766
Closes #6767

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6768)

jonthysell commented:
This might be a strictly a "nice to have" for 0.63. However it does resolve the issue that if you try to create a new project and immediately try to build in VS, that the autolink check fails, even though the only differences in the files are trivial.

asklar commented on packages/@react-native-windows/cli/src/generator-common/index.ts:
```diff
@@ -43,6 +43,22 @@ export function resolveContents(
+    // CRLF file, make sure multiline replacements are also CRLF
+    for (const key of Object.keys(replacements)) {
+      if (typeof replacements[key] === 'string') {
+        replacements[key] = replacements[key].replace(/([^\r]?)\n/g, '$1\r\n');
```
if it does find a \r, then $1 will be \r, meaning it will replace it with \r\r\n ?

NickGerleman commented:
Does this also fix https://github.com/microsoft/react-native-windows/issues/5932 ? This seems okay to backport to me. The telem changes were backported to 0.63, so the error on first build is a regression from earlier 0.63 builds.

NickGerleman commented on packages/@react-native-windows/cli/src/generator-common/index.ts:
```diff
@@ -43,6 +43,22 @@ export function resolveContents(
+    // CRLF file, make sure multiline replacements are also CRLF
+    for (const key of Object.keys(replacements)) {
+      if (typeof replacements[key] === 'string') {
+        replacements[key] = replacements[key].replace(/([^\r]?)\n/g, '$1\r\n');
```
This wouldn't do quite the right thing for the case of multiple new lines ("/n/n" would become "/n/r/n").

I think this is correct with look behind. This is what I ended up doing for CRLF normalization in override tooling e.g. "replace(/(?<!\r)\n/g, '\r\n')"

NickGerleman replied:
UTs are nice for these sorts of easy input to output algorithms (e.g. line space normalization). I had the exact same issue for my own implementation before adding tests for the interesting cases 🙂.

NickGerleman replied:
Super quick example you can run on a console (using visible chars instead of escape chars)
![image](https://user-images.githubusercontent.com/835219/102553184-523dcd80-4077-11eb-9efa-990c46b275ea.png)

NickGerleman commented on packages/@react-native-windows/cli/src/generator-windows/index.ts:
```diff
@@ -245,7 +245,7 @@ export async function copyProjectTemplateAndReplace(
   const templateVars: Record<string, any> = {
     useMustache: true,
-    regExpPatternsToRemove: ['//\\sclang-format\\s(on|off)\\s'],
+    regExpPatternsToRemove: ['//\\sclang-format\\s(on|off)\\s*'], // CODESYNC: vnext\local-cli\runWindows\utils\autolink.js
```
You could also just disable clang-format in the whole aggregate template directory by nesting a config file for it, instead having the template file disable it. We used to do that for C++ overrides.

jonthysell commented:
@NickGerleman @asklar Okay, I've:

1. Updated the normalization regex
2. Added a `.clang-format` file to exclude the template directory
3. Removed the unnecessary `//clang-format` guards from the template files.

Note that we want to **keep** the `//clang-format` guards in the generated autolinked files, so if the end customer starts using clang, those files remain ignored and therefore autolinking can still do a straight compare. We don't want to get into a situation where autolinking and clang are fighting over those files, and I don't want to make the file compare more complicated, as it is already the slowest part of autolinking.
